bindings/rust/s2n-tls-hyper/src/connector.rs

```diff
@@ -0,0 +1,180 @@
+{
+    http: T,
+    conn_builder: B,
+}
+
+impl<B> HttpsConnector<HttpConnector, B>
+where
+    B: connection::Builder,
+    <B as connection::Builder>::Output: Unpin,
+{
+    /// Creates a new `Builder` used to create an `HttpsConnector`.
+    ///
+    /// This builder is created using the default hyper `HttpConnector`. To use a custom HTTP
+    /// connector, use `HttpsConnector::builder_with_http()`.
+    pub fn builder(conn_builder: B) -> Builder<HttpConnector, B> {
+        let mut http = HttpConnector::new();
+        http.enforce_http(false);
+
+        Builder::new(Self { http, conn_builder })
+    }
+}
+
+impl<T, B> HttpsConnector<T, B>
+where
+    B: connection::Builder,
+    <B as connection::Builder>::Output: Unpin,
+{
+    /// Creates a new `Builder` used to create an `HttpsConnector`.
+    pub fn builder_with_http(http: T, conn_builder: B) -> Builder<T, B> {
+        Builder::new(Self { http, conn_builder })
+    }
+}
+
+impl<T, B> Service<Uri> for HttpsConnector<T, B>
+where
+    T: Service<Uri>,
+    T::Response: Read + Write + Connection + Unpin + Send + 'static,
+    T::Future: Send + 'static,
+    T::Error: Into<BoxError>,
+    B: connection::Builder + Send + Sync + 'static,
+    <B as connection::Builder>::Output: Unpin + Send,
+{
+    type Response = MaybeHttpsStream<T::Response, B>;
+    type Error = BoxError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<MaybeHttpsStream<T::Response, B>, BoxError>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self.http.poll_ready(cx) {
+            Poll::Ready(Ok(())) => Poll::Ready(Ok(())),
+            Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    fn call(&mut self, req: Uri) -> Self::Future {
+        if req.scheme() == Some(&http::uri::Scheme::HTTP) {
+            return Box::pin(async move { Err(UnsupportedScheme.into()) });
+        }
+
+        let builder = self.conn_builder.clone();
+        let host = req.host().unwrap_or("").to_owned();
+        let call = self.http.call(req);
+        Box::pin(async move {
+            let tcp = call.await.map_err(Into::into)?;
+            let tcp = TokioIo::new(tcp);
+
+            let connector = TlsConnector::new(builder);
+            let tls = connector.connect(&host, tcp).await?;
+
+            Ok(MaybeHttpsStream::Https(TokioIo::new(tls)))
+        })
+    }
+}
+
+pub struct Builder<T, B>
+where
+    B: connection::Builder,
+    <B as connection::Builder>::Output: Unpin,
+{
+    connector: HttpsConnector<T, B>,
+}
+
+impl<T, B> Builder<T, B>
+where
+    B: connection::Builder,
+    <B as connection::Builder>::Output: Unpin,
+{
+    pub fn new(connector: HttpsConnector<T, B>) -> Self {
+        Self { connector }
+    }
+
+    pub fn build(self) -> HttpsConnector<T, B> {
+        self.connector
+    }
+}
+
+#[derive(Debug)]
+struct UnsupportedScheme;
```

Comment:
This error feels a little odd, as does returning a single `BoxError` from the `call` function.

I expect that customers would generally want to handle different kinds of errors differently.
Returning a `Box<dyn std::error::Error>` makes it hard for customers to match on specific errors types, and similarly returning individual error structs like UnsupportedScheme means that a customer can't really use pattern matching to differentiate between them.

If you just want to move ahead with sketching out the initial shape of things I'm fine with that, but then maybe we could add a tracking issued for "topics to audit before s2n-tls-hyper 0.0.1 release".